exercises/practice/ocr-numbers/t/ocr-numbers.rakutest

```diff
@@ -0,0 +1,125 @@
+  { ascii-to-digits(" _ \n| |\n   ") },
+  Exception,
+  message => 'Number of input lines is not a multiple of four',
+  "Input with a number of lines that is not a multiple of four raises an error",
+); # end: cb19b733-4e36-4cf9-a4a1-6e6aac808b9a
+
+throws-like( # begin: 235f7bd1-991b-4587-98d4-84206eec4cc6
+  { ascii-to-digits("    \n   |\n   |\n    ") },
+  Exception,
+  message => 'Number of input columns is not a multiple of three',
+  "Input with a number of columns that is not a multiple of three raises an error",
+); # end: 235f7bd1-991b-4587-98d4-84206eec4cc6
+
+cmp-ok( # begin: 4a841794-73c9-4da9-a779-1f9837faff66
+    ascii-to-digits("       _     _        _  _ \n  |  || |  || |  |  || || |\n  |  ||_|  ||_|  |  ||_||_|\n                           "),
+    'eq',
+    '110101100',
+    "Recognizes 110101100",
+); # end: 4a841794-73c9-4da9-a779-1f9837faff66
+
+cmp-ok( # begin: 70c338f9-85b1-4296-a3a8-122901cdfde8
+    ascii-to-digits("       _     _           _ \n  |  || |  || |     || || |\n  |  | _|  ||_|  |  ||_||_|\n                           "),
+    'eq',
+    '11?10?1?0',
+    "Garbled numbers in a string are replaced with ?",
+); # end: 70c338f9-85b1-4296-a3a8-122901cdfde8
+
+cmp-ok( # begin: ea494ff4-3610-44d7-ab7e-72fdef0e0802
+    ascii-to-digits(" _ \n _|\n|_ \n   "),
+    'eq',
+    '2',
+    "Recognizes 2",
+); # end: ea494ff4-3610-44d7-ab7e-72fdef0e0802
+
+cmp-ok( # begin: 1acd2c00-412b-4268-93c2-bd7ff8e05a2c
+    ascii-to-digits(" _ \n _|\n _|\n   "),
+    'eq',
+    '3',
+    "Recognizes 3",
+); # end: 1acd2c00-412b-4268-93c2-bd7ff8e05a2c
+
+cmp-ok( # begin: eaec6a15-be17-4b6d-b895-596fae5d1329
+    ascii-to-digits("   \n|_|\n  |\n   "),
+    'eq',
+    '4',
+    "Recognizes 4",
+); # end: eaec6a15-be17-4b6d-b895-596fae5d1329
+
+cmp-ok( # begin: 440f397a-f046-4243-a6ca-81ab5406c56e
+    ascii-to-digits(" _ \n|_ \n _|\n   "),
+    'eq',
+    '5',
+    "Recognizes 5",
+); # end: 440f397a-f046-4243-a6ca-81ab5406c56e
+
+cmp-ok( # begin: f4c9cf6a-f1e2-4878-bfc3-9b85b657caa0
+    ascii-to-digits(" _ \n|_ \n|_|\n   "),
+    'eq',
+    '6',
+    "Recognizes 6",
+); # end: f4c9cf6a-f1e2-4878-bfc3-9b85b657caa0
+
+cmp-ok( # begin: e24ebf80-c611-41bb-a25a-ac2c0f232df5
+    ascii-to-digits(" _ \n  |\n  |\n   "),
+    'eq',
+    '7',
+    "Recognizes 7",
+); # end: e24ebf80-c611-41bb-a25a-ac2c0f232df5
+
+cmp-ok( # begin: b79cad4f-e264-4818-9d9e-77766792e233
+    ascii-to-digits(" _ \n|_|\n|_|\n   "),
+    'eq',
+    '8',
+    "Recognizes 8",
+); # end: b79cad4f-e264-4818-9d9e-77766792e233
+
+cmp-ok( # begin: 5efc9cfc-9227-4688-b77d-845049299e66
+    ascii-to-digits(" _ \n|_|\n _|\n   "),
+    'eq',
+    '9',
+    "Recognizes 9",
+); # end: 5efc9cfc-9227-4688-b77d-845049299e66
+
+cmp-ok( # begin: f60cb04a-42be-494e-a535-3451c8e097a4
+    ascii-to-digits("    _  _     _  _  _  _  _  _ \n  | _| _||_||_ |_   ||_||_|| |\n  ||_  _|  | _||_|  ||_| _||_|\n                              "),
```

Comment:
Especially for strings like this.  However, I wonder if the spacing that is required would be hard to enforce, or easily deleted in maintenance.

That said, does it make sense that the last line in digit is defined as a line, meaning that it requires an EOL (`\n`) to be a valid line?